### ISSUE TITLE
Assertion error in Flex Attention backward pass when indexing a parameter

### ISSUE THREAD

**Issue by otoomey** [pytorch/pytorch#146896]
### 🐛 Describe the bug

Flex Attention raises an assertion error during the backward pass if the `score_mod` implementation indexes a unit sized dimension.

For instance:
```python
import torch
from torch.nn.attention.flex_attention import flex_attention

MAXLEN = 2
embeddings_table = nn.Parameter(torch.Tensor(2 * MAXLEN, 1)) # <-- note unit sized last dimension

query = torch.randn(1, 1, 1, 32)
key = torch.randn(1, 1, 1, 32)
value = torch.randn(1, 1, 1, 32)
dt = torch.randn(1, 1, 1)

def rpe(score, _1, _2, q_idx, kv_idx):
    delta = q_idx - kv_idx
    delta = torch.clamp(delta, -MAXLEN, MAXLEN - 1)
    delta += MAXLEN
    return score + embeddings_table[delta.int()][0] # <-- note the [0] here

attn_output = flex_attention(
    query=query,
    key=key,
    value=value,
    score_mod=rpe,
)

attn_output.sum().backward()
```
An assertion error is raised on the last line:
```
File /.../.venv/lib/python3.11/site-packages/torch/_dynamo/_trace_wrapped_higher_order_op.py:84, in _(info, indims, shape, indices, value)
     81         expanded_indices.append(idx.expand(value.shape))
     82     else:
     83         # the index is being part of the vmap batch, it should be the same size as val
---> 84         assert idx.shape == value.shape
     85         expanded_indices.append(idx)
     87 out = torch.ops.flex_lib.zeros_and_scatter(
     88     shape,
     89     expanded_indices,
     90     value,
     91 )

AssertionError:
```
If the unit sized dimension is removed from `embeddings_table` the code works as expected.

### Versions

```
Collecting environment information...
PyTorch version: 2.7.0.dev20250211+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: AlmaLinux release 8.10 (Cerulean Leopard) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-22)
Clang version: 17.0.6 (Red Hat 17.0.6-1.module_el8.10.0+3757+fc27b834)
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.11.3 (main, Apr 19 2023, 23:54:32) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.18.0-553.8.1.el8_10.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 530.30.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              24
On-line CPU(s) list: 0-23
Thread(s) per core:  2
Core(s) per socket:  12
Socket(s):           1
NUMA node(s):        1
Vendor ID:           AuthenticAMD
CPU family:          25
Model:               97
Model name:          AMD Ryzen 9 7900X 12-Core Processor
Stepping:            2
CPU MHz:             545.000
CPU max MHz:         5733.0000
CPU min MHz:         545.0000
BogoMIPS:            9399.80
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            32768K
NUMA node0 CPU(s):   0-23
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 clzero irperf xsaveerptr wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid overflow_recov succor smca fsrm flush_l1d

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu11==2.20.5
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.16.2
[pip3] pytorch-triton==3.2.0+git4b3bb1f8
[pip3] torch==2.7.0.dev20250211+cu124
[pip3] torch-tb-profiler==0.4.3
[pip3] torchaudio==2.6.0.dev20250211+cu124
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.6.1
[pip3] torchvision==0.22.0.dev20250211+cu124
[pip3] triton==3.1.0
[conda] blas                      1.0                         mkl  
[conda] cuda-cudart               11.8.89                       0    nvidia
[conda] cuda-cupti                11.8.87                       0    nvidia
[conda] cuda-libraries            11.8.0                        0    nvidia
[conda] cuda-nvrtc                11.8.89                       0    nvidia
[conda] cuda-nvtx                 11.8.86                       0    nvidia
[conda] cuda-runtime              11.8.0                        0    nvidia
[conda] cudatoolkit               11.8.0               h6a678d5_0  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libcublas                 11.11.3.6                     0    nvidia
[conda] libcufft                  10.9.0.58                     0    nvidia
[conda] libcurand                 10.3.3.53                     0    nvidia
[conda] libcusolver               11.4.1.48                     0    nvidia
[conda] libcusparse               11.7.5.86                     0    nvidia
[conda] mkl                       2023.1.0         h6d00ec8_46342  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.6           py311ha02d727_1  
[conda] mkl_random                1.2.2           py311ha02d727_1  
[conda] numpy                     1.24.3          py311h08b1b3b_1  
[conda] numpy-base                1.24.3          py311hf175353_1  
[conda] numpydoc                  1.5.0           py311h06a4308_0  
[conda] pytorch                   2.0.1           py3.11_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.2               py311_cu118    pytorch
[conda] torchtriton               2.0.0                     py311    pytorch
[conda] torchvision               0.15.2              py311_cu118    pytorch
```

**Comment by CodeCreator** [comment]
+1 also observing the same behavior when adding a `attn_outputs.sum().backward()` to the code snippet from #145460

**Comment by SirWyver** [comment]
I've encountered the same issue - I'm wondering if this is just a bug or  rather something that takes longer to support?

**Comment by drisspg** [comment]
Okay yeah there is like 100 things going on here. 

1. One thing is that the CPU path especially for training is much less tested, and I think since your repro uses cpu tensors there is a bug related to that implementation @jianan-gu if we are somehow allowing training on for CPU tensors w/ compile we need to be sure to error

With that in mind I updated the repro to use the cuda backend
```Py
    MAXLEN = 2
    embeddings_table = nn.Parameter(torch.randn((2 * MAXLEN, 1), device="cuda")) # <-- note unit sized last dimension
    query = torch.randn(1, 1, 1, 32, device="cuda")
    key = torch.randn(1, 1, 1, 32, device="cuda")
    value = torch.randn(1, 1, 1, 32, device="cuda")
    dt = torch.randn(1, 1, 1, device="cuda")

    def rpe(score, _1, _2, q_idx, kv_idx):
        delta = q_idx - kv_idx
        delta = torch.clamp(delta, -MAXLEN, MAXLEN - 1)
        delta += MAXLEN
        lookup = embeddings_table[:, 0][delta.to(torch.int32)] # <-- note the [0] here
        return score + lookup

    attn_output = flex_attention(
        query=query,
        key=key,
        value=value,
        score_mod=rpe,
    )

    attn_output.sum().backward()
``` 

You can see that the embedding table index is alot different then what you have. One of the constraints right now for Trainable biases is that we only support 1 index to the object, I have a nice error message (that didnt trigger) here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/kernel/flex_attention.py#L2292-L2303

I had plans to potentially rewrite this `embeddings_table[delta.int()][0]` fx graph to use 1 call if possible. 

So the question is why didn't we get to this error message? And there are number of things but the main one is that 
the naive update 

` embeddings_table[delta.to(torch.int32), 0] ` actually produces a scalar that implicitly syncs to CPU and FlexAttention does not support that (since it cant be 1 kernel). The above was 1 weird slicing + indexing permutation I found that works but yeah this is frustrating and we need to make this better

**Comment by aorenste** [comment]
Still reproduces on PyTorch 2.13.0.dev20260510+cpu. Minimized repro (down to ~7 non-import lines):

```python
import torch
from torch.nn.attention.flex_attention import flex_attention

embeddings_table = torch.randn(4, 1, requires_grad=True)
qkv = torch.randn(1, 1, 1, 8)

def rpe(score, _1, _2, q_idx, kv_idx):
    return score + embeddings_table[q_idx - kv_idx + 2][0]

flex_attention(qkv, qkv, qkv, score_mod=rpe).sum().backward()
```

The assertion fires in `torch._dynamo._trace_wrapped_higher_order_op` during the backward pass when `zeros_and_scatter` is called for the score_mod's autograd. The trigger seems to be the combination of (a) indexing a 2-D parameter table with a tensor index inside `score_mod` and then taking `[0]` to drop the unit-sized trailing dim, and (b) requiring backward through that parameter. Replacing the `[q_idx - kv_idx + 2][0]` pattern with `[q_idx]` alone (no subtraction-based delta and no trailing `[0]`) makes the bug disappear, so both the broadcasted index shape and the unit-dim trailing slice appear to be necessary.

_(Analysis done by Claude.)_
